platform/orion/core/generic/delivery/delivery.go

```diff
@@ -52,51 +52,54 @@ type DeliverStream interface {
 }
 
 type delivery struct {
-	ctx                 context.Context
 	sp                  view2.ServiceProvider
 	network             Network
 	waitForEventTimeout time.Duration
 	callback            Callback
 	vault               Vault
 	me                  string
 	networkName         string
+	stop                chan bool
 }
 
 func New(
-	ctx context.Context,
 	sp view2.ServiceProvider,
 	network Network,
 	callback Callback,
 	vault Vault,
 	waitForEventTimeout time.Duration,
 ) (*delivery, error) {
-	if ctx == nil {
-		ctx = context.Background()
-	}
 	d := &delivery{
-		ctx:                 ctx,
 		sp:                  sp,
 		network:             network,
 		waitForEventTimeout: waitForEventTimeout,
 		callback:            callback,
 		vault:               vault,
 		me:                  network.IdentityManager().Me(),
 		networkName:         network.Name(),
+		stop:                make(chan bool),
 	}
 	return d, nil
 }
 
-// Start runs the delivery service in a goroutine
-func (d *delivery) Start() {
-	go d.Run()
+// StartDelivery runs the delivery service in a goroutine
+func (d *delivery) StartDelivery(ctx context.Context) error {
+	go d.Run(ctx)
+	return nil
 }
 
-func (d *delivery) Run() error {
+func (d *delivery) Run(ctx context.Context) error {
+	if ctx == nil {
+		ctx = context.Background()
+	}
 	var df DeliverStream
 	var err error
 	for {
 		select {
-		case <-d.ctx.Done():
+		case <-d.stop:
+			// Time to cancel
+			return errors.New("stop")
+		case <-ctx.Done():
 			// Time to cancel
 			return errors.New("context done")
```

Comment:
why would you return an error when you send a "graceful" stop?

Reply:
In line `155` there the we also stop without an error.

Reply:
addressed